src/main/java/net/dv8tion/jda/core/requests/WebSocketClient.java

```diff
@@ -1006,32 +1008,43 @@ protected void handleEvent(JSONObject raw)
         }
     }
 
+    public boolean onBufferMessage(byte[] binary) throws IOException
+    {
+        if (binary.length >= 4 && getInt(binary, binary.length - 4) == ZLIB_SUFFIX)
+            return true;
+        if (readBuffer != null)
+            readBuffer.write(binary);
+        else
+            allocateBuffer(binary);
+
+        return false;
+    }
+
     @Override
-    public void onBinaryMessage(WebSocket websocket, byte[] binary) throws IOException
+    public void onBinaryMessage(WebSocket websocket, byte[] binary) throws IOException, DataFormatException
     {
-        readBuffer.write(binary);
-        //check whether the transition is finished or not
-        if (readBuffer.size() < 4)
-            return;
-        final int suffix = getInt(binary, binary.length - 4);
-        if (suffix != ZLIB_SUFFIX)
+        if (!onBufferMessage(binary))
             return;
         //Thanks to ShadowLordAlpha and Shredder121 for code and debugging.
         //Get the compressed message and inflate it
-        ByteArrayOutputStream out = new ByteArrayOutputStream(readBuffer.size() * 2);
+        final int size = readBuffer != null ? readBuffer.size() : binary.length;
+        ByteArrayOutputStream out = new ByteArrayOutputStream(size * 2);
         try (InflaterOutputStream decompressor = new InflaterOutputStream(out, zlibContext))
         {
-            readBuffer.writeTo(decompressor);
+            if (readBuffer != null)
+                readBuffer.writeTo(decompressor);
+            else
+                decompressor.write(binary);
```

Comment:
Based on this code, it looks like you're relying onBufferMessage to return true when it has found the end of the ZLIB packet, however, take the following situation:
onBinaryMessage receives a partial zlib packet. We determine it is partial in onBufferMessage, place the partial in a buffer, and return false from onBufferMessage.
onBinaryMessage receives a second zlib packet, this one containing the rest of the partial packet AND the zlib suffix. onBufferMessage will return true without writting the data to the buffer.

Since onBufferMessage returned true, we attempt to decompress the data, however, the readBuffer will only have the partial packet data from the first packet.

Am I missing something here?